designs/v1beta1-api.md

```diff
@@ -0,0 +1,341 @@
+spec:
+  amiFamily: AL2
+  amiSelector:
+    - tags: 
+        key: value
+    - id: abc-123
+    - name: foo
+      owner: amazon
+    - ssm: "/my/custom/parameter"
+  subnetSelector:
+    - tags:
+        compute.k8s.aws/discovery: cluster-name
+    - id: subnet-1234
+  securityGroupSelector:
+    - tags:
+        compute.k8s.aws/discovery: cluster-name
+    - name: default-security-group
+  role: karpenter-node-role
+  userData: |
+    echo "this is custom user data"
+  tags:
+    custom-tag: custom-value
+  metadataOptions:
+    httpEndpoint: enabled
+    httpProtocolIPv6: disabled
+    httpPutResponseHopLimit: 2
+    httpTokens: required
+  blockDeviceMappings:
+    - deviceName: /dev/xvda
+      ebs:
+        volumeSize: 20Gi
+        volumeType: gp3
+        encrypted: true
+  detailedMonitoring: true
+status:
+  subnets:
+    - id: subnet-0a462d98193ff9fac
+      zone: us-east-2b
+    - id: subnet-0322dfafd76a609b6
+      zone: us-east-2c
+    - id: subnet-0727ef01daf4ac9fe
+      zone: us-east-2b
+    - id: subnet-00c99aeafe2a70304
+      zone: us-east-2a
+    - id: subnet-023b232fd5eb0028e
+      zone: us-east-2c
+    - id: subnet-03941e7ad6afeaa72
+      zone: us-east-2a
+  securityGroups:
+    - id: sg-041513b454818610b
+      name: ClusterSharedNodeSecurityGroup
+    - id: sg-0286715698b894bca
+      name: ControlPlaneSecurityGroup-1AQ073TSAAPW
+  amis:
+    - id: ami-05a05e85b17bb60d7
+      name: amazon-eks-node-1.24-v20230703
+      requirements:
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: DoesNotExist
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - amd64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: DoesNotExist
+    - id: ami-0d849ef1e65103147
+      name: amazon-eks-gpu-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - amd64
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: Exists
+    - id: ami-0d849ef1e65103147
+      name: amazon-eks-gpu-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+           - amd64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: Exists
+    - id: ami-0c3487f30d003deb3
+      name: amazon-eks-arm64-node-1.24-v20230703
+      requirements:
+        - key: kubernetes.io/arch
+          operator: In
+          values:
+            - arm64
+        - key: karpenter.k8s.aws/instance-gpu-count
+          operator: DoesNotExist
+        - key: karpenter.k8s.aws/instance-accelerator-count
+          operator: DoesNotExist
+```
+
+### `karpenter.sh/NodeClaim`
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodeClaim
+metadata:
+  name: default
+  labels:
+    billing-team: my-team
+  annotations:
+    example.com/owner: "my-team"
+spec:
+  nodeClass:
+    name: default
+    kind: NodeClass
```

Comment:
It's slightly more accurate to specify the API path (eg `nodeclasses`) than the kind (kinds can clash; API paths never do)

Reply:
Pathological case is something like:
- `/apis/cloud-provider.example/v1beta1/providernodeclasses/foo`
- `/apis/cloud-provider.example/v1beta1/nodeclasses/foo`

both say that their kind is NodeClass (yes, I know, don't do this).

Using `providernodeclasses` rather than NodeClass would  avoid that risk.

Reply:
Huh. I didn't realize that you could create a Kind that had a different path than the actual kind name. Regardless if that's the case, I don't think that people should be doing this and (if someone is doing this) we shouldn't try to build support around our API to respect it.

Reply:
SIG Auth has to think about the risks here. For Karpenter I think it's fine. Anyway, I wanted to make sure we made a decision with good context.